File: src/Airport.fs
module Airport

open FSharp.Data
open System
open System.IO
open System.Net
open Util
open Util.Result

type Airports = CsvProvider<Schema = ",Ident (string),Type (string),Name (string),Lat (float),Lon (float),,Continent (string)",
                            CacheRows = false, HasHeaders = false>

type Type =
    | Closed
    | Heliport
    | Small
    | Medium
    | Large

type Airport = {
    Name      : string
    ICAO      : string
    Coord     : Coord.T
    Type      : Type
    Continent : string
}

let distance from to' = Coord.distance from.Coord to'.Coord
let findFromICAO icao = Array.tryFind (fun a -> a.ICAO = icao)

let timeBetween mach from to' =
    let metersPerSecond = 343.
    let distance = distance from to'
    TimeSpan.FromSeconds(distance / (mach * metersPerSecond))

let loadAll (path : string) =
    let readType = function
        | "closed"         -> Closed
        | "heliport"       -> Heliport
        | "small_airport"  -> Small
        | "medium_airport" -> Medium
        | "large_airport"  -> Large
        | _                -> Closed

    let createInfo (x : Airports.Row) = {
        Name  = x.Name
        ICAO  = x.Ident
        Coord = Coord.create
                    (double x.Lat)
                    (double x.Lon)
        Type      = readType x.Type
        Continent = x.Continent
    }

    (Airports.Load path).Rows
    |> Seq.tail // Skip the row containing the headers
    |> Seq.map createInfo

module DataFile =
    let name = "airports.csv"
    let path = Util.localPath name

    let isOld path =
        let info = FileInfo path
        not info.Exists || (DateTime.UtcNow - info.LastWriteTimeUtc).TotalDays > 30.

    let update () =
        try
            use wc = new WebClient()
            wc.DownloadFile(
                sprintf "http://ourairports.com/data/%s" name,
                path)
            Success ()
        with
        | ex -> Failure ex.Message

    let verifyAndUpdate () =
        match isOld path with
        | true ->
            printfn "Airport data out of date. Updating.."
            update ()
        | false -> Success ()

File: src/airport.rs
use std::ascii::AsciiExt;
use ::rocket::http::RawStr;
use ::rocket::request::FromFormValue;

#[derive(Debug, Serialize)]
pub struct ICAO(pub String);

impl<'v> FromFormValue<'v> for ICAO {
    type Error = &'v RawStr;

    fn from_form_value(form_value: &'v RawStr) -> Result<ICAO, &'v RawStr> {
        if form_value != "" {
            Ok(ICAO(form_value.as_str().into()))
        } else {
            Err(form_value)
        }
    }
}

#[derive(Debug)]
pub enum Size {
    Heliport,
    Small,
    Medium,
    Large
}

impl<'v> FromFormValue<'v> for Size {
    type Error = &'v RawStr;

    fn from_form_value(form_value: &'v RawStr) -> Result<Size, &'v RawStr> {
        let value = form_value.as_str().to_ascii_lowercase();

        match value.as_str() {
            "heliport" => Ok(Size::Heliport),
            "small"    => Ok(Size::Small),
            "medium"   => Ok(Size::Medium),
            "large"    => Ok(Size::Large),
            _          => Err(form_value),
        }
    }
}

#[derive(Debug, Serialize)]
pub struct LatLon {
    pub lat: f64,
    pub lon: f64,
}

impl LatLon {
    pub fn new(lat: f64, lon: f64) -> LatLon {
        LatLon { lat, lon }
    }
}

#[derive(Debug, Serialize)]
pub struct Airport {
    pub pos:  LatLon,
    pub icao: ICAO,
}